File: 01Linkedlist.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstdlib>
/*
Insertion Sort. Given a file contains a list of English words,
your task is to construct a linked list (in ascending order) for the words.
Your linked list must have a dummy node that is pointed by the listHead.
For example,
Insert (you):
listHead --> (dummy, you) --> (you,  null)
*/

using namespace std;

class linkedList{
//to create a class node
    class listNode{
    public:
        string data;
        listNode *next;

        listNode(){
            next = NULL;
        }

        listNode(string str){
            data = str;
            next = NULL;
        }

    };

public:
    listNode *listHead;

    linkedList(string ele){
        listHead = new listNode(ele);
    }

    bool isEmpty(){
        if(listHead == NULL){
            return true;
        }
        return false;
    }

    void listInsertOrdered(string data){
//to insert a data according to alphabetical order
        listNode *curr = listHead->next;
        listNode **target = &listHead->next;
        listNode *newNode = new listNode(data);

        if(curr == NULL){
            newNode->next = curr;
            listHead->next = newNode;
        }else{
            while((curr != NULL)&&(curr->data.compare(data)< 0)){
            target = &curr->next;
            curr = curr ->next;
            }
            *target = newNode;
            newNode->next = curr;
        }
        return;
    }

    void listInsert(string data){
//just to insert a data
        listNode *po = listHead;
        listNode *node = new listNode(data);

        if(po->next == NULL){
            node->next = listHead->next;
            listHead->next = node;
        }else{
            while(po->next != NULL){
                po = po->next;
            }
            node->next = po->next;
            po->next = node;
        }
    }

    void listDelete(string data){
        listNode *curr = listHead;
        listNode **prev = &listHead;

        while((curr != NULL)&&(data != curr->data)){
            prev = &curr->next;
            curr = curr->next;
        }

        if(curr != NULL){
            *prev = curr->next;
            delete curr;
        }
        return;
    }

    void printList(string ele, ofstream &output){

		listNode *curr = listHead;
        output << "Insert" << " (" + ele + "):\n";
		output << "listHead-->";

        while(true){
            if(curr->next == NULL){
                output << "(" << curr->data << ", NULL)\n" << endl;
                return;
            }else{
                output << "(" << curr->data << ","<< curr->next->data<<")-->";
            }
            curr = curr->next;
		}
    }

};

int main(int argc, char *argv[]) {
//code::block to set up the arguments: project->set program's argument (argv[1],argv[2])
    ifstream inputEle(argv[1]);
    ofstream outputEle(argv[2]);

    linkedList theList("dummy");

    string ele;
    int icount = 0;

    bool bo = theList.isEmpty();
//to insert 15 nodes.
    while(inputEle >> ele && icount <15){
        theList.listInsertOrdered(ele);
        theList.printList(ele, outputEle);
        icount++;
    }

    return 0;
}

File: 01linkedList.java
import java.io.*;
import java.util.*;

public class linkedList {

	static class listNode{
		public String data;
		public listNode next;
		
		listNode(){
			next = null;
		}
		
		listNode(String ele){
			data = ele;
			next = null;
		}
		
	}

	public listNode listHead = new listNode();
	
	public linkedList(String ele) {
		listHead.data = ele;
	}
	
	public boolean isEmpty() {
        if(listHead == null){
            return true;
        }
        return false;
		
	}
	
	public void listInsert(String ele) {
        listNode curr = listHead.next;
        listNode prev = listHead;
        listNode newNode = new listNode(ele);
        
        if(curr == null){
            newNode.next = curr;
            listHead.next = newNode;
        }else{
            while(curr.next != null) {
            	if(newNode.data.compareToIgnoreCase(curr.data) < 0){
            		newNode.next = curr;
            		prev.next = newNode;
            		return;
            	}else {
            		curr = curr.next;
            		prev = prev.next;
            	}
            }

            if(newNode.data.compareToIgnoreCase(curr.data) < 0) {
	            newNode.next = curr;
	            prev.next = newNode;     
            }else {
            	curr.next = newNode;
            }
        }
        return;
	}
	
	public void listDelete(String ele) {

        listNode curr = listHead.next;
        listNode prev = listHead;

        while((curr != null)&&(curr.data.equals(ele))){
            prev = prev.next;
            curr = curr.next;
        }

        if(curr != null){
            prev.next = curr.next;
            curr.next = null;
        }
        return;
	}
	
	public void printList(String ele, PrintWriter pw) {
		listNode curr = listHead;
		pw.println("Insert (" + ele + "):\n");
		pw.print("listHead-->");

        while(true){
            if(curr.next == null){
            	pw.println("("+ curr.data + ", NULL)");
                return;
            }else{
            	pw.print("(" + curr.data + "," + curr.next.data + ")-->");
            }
            curr = curr.next;
		}
	}

	
	public static void main(String[] args){
		
	linkedList ls = new linkedList("dummy");
		try {
		File myFileR = new File(args[0]);
		FileReader fr = new FileReader(myFileR);
	     
	    FileWriter fw = new FileWriter(args[1]);
	    PrintWriter pw = new PrintWriter(fw);

		Scanner myScan = new Scanner(fr);
		int icount = 0;
		
		while(myScan.hasNext() && icount < 15) {
			String ele = myScan.next();		
			ls.listInsert(ele);
			System.out.println(ele);
			ls.printList(ele, pw);		
			icount++;
		}
		
		fr.close();
		myScan.close();
		pw.close();
		}catch(Exception e) {}
		
	}

}